Getting Started docs: Add `caching and revalidating` page (#73437)
vercel/next.js: docs/01-app/01-getting-started/07-caching-and-revalidating.mdx

---
title: How to cache and revalidate components and functions
nav_title: Caching and Revalidating
description: Learn how to cache and revalidate components and functions in your Next.js application.
related:
  title: API Reference
  description: Learn more about the features mentioned in this page by reading the API Reference.
  links:
    - app/api-reference/config/next-config-js/dynamicIO
    - app/api-reference/directives/use-cache
    - app/api-reference/functions/cacheLife
    - app/api-reference/functions/cacheTag
    - app/api-reference/functions/revalidateTag
    - app/api-reference/functions/revalidatePath
---

> **Warning:** The content below assumes the [`dynamicIO` config option](/docs/app/api-reference/config/next-config-js/dynamicIO) is enabled in your application. This option was introduced in Next.js 15 canary.

## Caching

In Next.js, you can cache the output of a [component](#components) or [function](#functions) to reduce the need to re-execute work for every user request.

> Caching is useful for data that doesn't change often or is shared across multiple users, as you can **prerender** at build time, and reuse the static content for all requests.

Reviewer: I wonder if we should talk about prerendering this high up, because you could also cache this at runtime (not prerendered) right?

Author: Update the wording slightly to remove "prerendering" and avoid confusion. I think the runtime / in-memory cache might be too detailed for this section. but we should ensure it's documented in the `use cache` page.

To mark a function or component as cacheable, you can use the [`'use cache'`](/docs/app/api-reference/directives/use-cache) directive.

### Components

You can add `'use cache'` to an **asynchronous** [Server Component](https://react.dev/reference/rsc/server-components) to cache the component's render output:

```tsx filename="app/blog/page.tsx" highlight={4} switcher
import { getPosts } from '@/app/lib/data'

export default async function Page() {
  'use cache'
  const posts = await getPosts()

  return (
    <ul>
      {posts.map((post) => (
        <li key={post.id}>{post.title}</li>
      ))}
    </ul>
  )
}
```

```js filename="app/blog/page.js" highlight={4} switcher
import { getPosts } from '@/app/lib/data'

export default async function Page() {
  'use cache'
  const posts = await getPosts()
  return (
    <ul>
      {posts.map((post) => (
        <li key={post.id}>{post.title}</li>
      ))}
    </ul>
  )
}
```

> Alternatively, you can add the `'use cache'` directive at the top of the file to mark all components in the file as cacheable:

Reviewer: Good clarifying question, maybe we list it or not, are there multiple cache entries for this? For the function and output? Or just the output?

```tsx filename="app/blog/page.tsx" highlight={1} switcher
'use cache'

import { getPosts } from '@/app/lib/data'

export default async function Page() {
  const posts = await getPosts()

  return <Posts posts={posts} />
}

async function Posts({ posts }) {
  return (
    <ul>
      {posts.map((post) => (
        <li key={post.id}>{post.title}</li>
      ))}
    </ul>
  )
}
```

```js filename="app/blog/page.js" highlight={1} switcher
'use cache'

import { getPosts } from '@/app/lib/data'

export default async function Page() {
  const posts = await getPosts()

  return <Posts posts={posts} />
}

async function Posts({ posts }) {
  return (
    <ul>
      {posts.map((post) => (
        <li key={post.id}>{post.title}</li>
      ))}
    </ul>
  )
}
```

### Functions

You can cache the return value of an **asynchronous** function, including data requests, by adding the `'use cache'` directive inside the function body:

```ts filename="app/lib/data.ts" switcher
export async function getPosts(slug: string) {
  'use cache'

>   const data = await fetch(`/api/posts/${slug}`)

Reviewer: Could we use api.vercel.app so they are real code snippets?

  return data.json()
}
```

```js filename="app/lib/data.js" switcher
export async function getPosts(slug) {
  'use cache'
  const data = await fetch(`/api/posts/${slug}`)
  return data.json()
}
```

> With the approach above, you can call the function throughout your application code and the same cache entry will be reused as long as the arguments are the same.

Reviewer: ```suggestion
You can then call the function throughout your application code and the same cache entry will be reused, as long as the arguments to the function (e.g. `slug`) are the same.
```

Alternatively, you can also cache all functions within a file by adding the `'use cache'` directive at the top of the file:

```ts filename="app/lib/data.ts" switcher
'use cache'

export async function getPosts() {
  const data = await fetch(`/api/posts`)
  return data.json()
}

export async function getPostBySlug(slug: string) {
  const data = await fetch(`/api/posts/${slug}`)
  return data.json()
}
```

```js filename="app/lib/data.js" switcher
'use cache'

export async function getPosts() {
  const data = await fetch(`/api/posts/`)
  return data.json()
}

export async function getPostBySlug(slug: string) {
  const data = await fetch(`/api/posts/${slug}`)
  return data.json()
}
```

> 

Reviewer: Maybe again: There will be two separate entries into the cache. One for `getPosts` and another for `getPostsBySlug`, where the cache will be invalidated if the `slug` argument changes.

Author: For the invalidated part, wouldn't it be two separate entries? A cache key is the function's unique identity + the input.

## Revalidating

> Revalidation allows you to update cached content without having to rebuild your entire application. It's useful for content that _sometimes_ changes, but would benefit from being cached to improve your application'sperformance.

Reviewer: ```suggestion
Revalidation allows you to update cached content without having to rebuild your entire application. It's useful for content that _sometimes_ changes, but would benefit from being cached to improve your application's performance.
```

- [On-demand](#on-demand-revalidation): Triggered by a specific event (e.g. a CMS webhook).

## Time-based Revalidation

You can use the [`cacheLife` function](/docs/app/api-reference/functions/cacheLife) to define a time interval for how long a cached value should remain stale before it's revalidated.

`cacheLife` comes with [default cache profiles](/docs/app/api-reference/functions/cacheLife#default-cache-profiles) such as `'hour'`, `'day'`, and `'week'`. These profiles can be [customized](/docs/app/api-reference/functions/cacheLife#custom-cache-profiles), if needed.

To use `cacheLife`, import it from `next/cache` and nest it within the **scope** of the `'use cache'` directive and the function. For example, to revalidate the blog page after one hour:

```tsx filename="app/blog/page.tsx" highlight={3,6} switcher
'use cache'o/>:.

import { cacheLife } from 'next/cache'

export default async function Page() {
  cacheLife('hour')
  return <Posts posts={posts} />
}
```

```jsx filename="app/blog/page.js" highlight={3,6} switcher
'use cache'

import { cacheLife } from 'next/cache'

export default async function Page() {
  cacheLife('hour')
  return <Posts posts={posts} />
}
```

> 

Reviewer: Maybe could stand for a collapsed "deep dive" that explains the stale-while-revalidate semantics, e.g. after one hour it's stale, updates in the background, and then the _next_ request sees the fresh content.

Reviewer 2: FYI I just landed here while searching about how stale-while-revalidate works with the `use cache` directive so it would be good to have it on the docs indeed.

```tsx filename="app/ui/user.tsx" switcher
import { cookies } from 'next/headers'

export async function User() {
  const store = await cookies()
  const { id } = store.get('user')
  const user = await fetch(`https://api.vercel.app/users/${id}`)
  return <>...</>
}
```

```jsx filename="app/ui/user.js" switcher
import { cookies } from 'next/headers'

export async function User() {
  const store = await cookies()
  const { id } = store.get('user')
  const user = await fetch(`https://api.vercel.app/users/${id}`)
  return <>...</>
}
```

To prevent the whole route from being blocked, you can wrap the component in a `Suspense` boundary. This means the `<Navigation>` component and a fallback will be sent to the client first, while the nested `<User>` component, which relies on a potentially blocking request, will be streamed when it's ready.

```tsx filename="app/ui/navigation.tsx" switcher
import { Suspense } from 'react'
import { Link } from 'next/link'
import { User } from '@/app/ui/user'

export default function Navigation() {
  return (
    <nav>
      <Link href="/">Home</Link>

>       <Suspense fallback={<div>Loading...</div>}>

Reviewer: We might want to recommend <Skeleton/>
